[3632] Create the new assets panel

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/3632

aachavarria committed:
assetPanel and Search service Changes

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -27,12 +44,104 @@ const translations = defineMessages({
-      </Typography>
+      <div className={classes.assetsPanelWrapper}>
+        <SearchBar onChange={() => {
+        }} keyword={''} classes={{ root: classes.search }}/>
```
Format

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -27,12 +44,104 @@ const translations = defineMessages({
+  }
+}));
+
+function AssetComponent(props: any) {
```
Use same card styles/structure from main search results page minus the checkbox and the options menu will have different options. Refactor search component as appropriately for card re-usage.

rart commented on ui/app/src/components/SearchBar.tsx:
```diff
@@ -84,7 +87,7 @@ export default function SearchBar(props: SearchBarProps) {
   const {formatMessage} = useIntl();
   return (
-    <div className={clsx(classes.search, focus && 'focus')}>
+    <div className={clsx(classes.search, focus && 'focus', props.classes?.root && props.classes.root)}>
```
`props.classes?.root` would be enough (no need for the && exp)

rart commented on ui/app/src/components/SearchBar.tsx:
```diff
@@ -76,6 +76,9 @@ interface SearchBarProps {
   backgroundColor?: string;
   placeholder?: string;
+  classes?: {
+    root: any;
```
All inner props of `classes` should also be optional (`?`)

aachavarria committed:
mising auth on gitForm privateKey

aachavarria committed:
MediaCard refactor, search changes to include menu and cardOptions

aachavarria committed:
mediaCard adding Avatar

aachavarria committed:
AssetsPanel changes

aachavarria committed:
adding new mime types, cleaning code

aachavarria committed:
adding Pagination

aachavarria committed:
improve on searchBar to support no keyword,
AssetsPanel search

aachavarria committed:
static pagination

aachavarria committed:
adding dnd to mediaCard

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

 Conflicts:
	ui/app/src/components/MediaCard.tsx
	ui/app/src/models/Search.ts

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

aachavarria committed:
adding dnd update field for images

rart commented on ui/app/.env:
```diff
@@ -1,3 +1,3 @@
-# HOST=127.0.0.1
+HOST=authoring.sample.com
```
Do not commit

rart replied:
DO NOT COMMIT

rart replied:
Check back here on Github's PR "Files changed" tab to make sure that you've successfully reverted and removed from commit.

rart commented on ui/app/src/components/SearchBar.tsx:
```diff
@@ -84,15 +87,15 @@ export default function SearchBar(props: SearchBarProps) {
         placeholder={placeholder || formatMessage(messages.placeholder)}
         autoFocus={autofocus}
-        value={keyword}
+        value={keyword !== 'undefined' && keyword}
```
did you mean `typeof keyword`?

You need to always supply a value or react will whine about switching input between controlled/uncontrolled

aachavarria replied:
react only complains when the value is null, right now is it's not throwing errors

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,207 @@
+    type: ASSET_DRAG_ENDED
+  });
+
+  setRequestForgeryToken();
```
Should be able to delete this now, no?

aachavarria replied:
I think no, if I delete it does not work

rart replied:
Please check why. This is set on the `CrafterCMSNextBridge.tsx`, which is always the root of the app. 

Are you on `/studio` when using the Webpack server?

aachavarria replied:
that should work when im on studio, but not on react app right?

rart replied:
I think it should work either way (that’s my point here)

aachavarria replied:
It is not sending the X-XSRF-TOKEN

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,207 @@
+    bottom: 0,
+    background: 'white',
+    color: 'black',
+    width: '239px',
```
There’s a DRAWER_WIDTH somewhere around there. In any case, preferably use a calc or percentage.

aachavarria committed:
Changes on update field

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

 Conflicts:
	ui/app/src/modules/Preview/PreviewConcierge.tsx
	ui/app/src/modules/System/Sites/Create/GitForm.tsx

aachavarria committed:
UPDATE_FIELD_VALUE_OPERATION

aachavarria committed:
adding import updateField

aachavarria committed:
creating actions, reducers, epics to fetchAssets

aachavarria committed:
changes to assets epic, reducer, adding resource to assetPanel

aachavarria committed:
refactor on useEntitySelectionResource, AssetsPanel

aachavarria committed:
epic fixes, resource suspense fixes

aachavarria committed:
fixes, PR Review

aachavarria committed:
adding support to videos on assets dnd

aachavarria committed:
adding scroll to assets

rart commented on ui/app/src/components/MediaCard.tsx:
```diff
@@ -119,68 +113,53 @@ const useStyles = makeStyles((theme: Theme) => ({
+  const { name, path, type } = item;
+  const isList = currentView === 'list';
 
   const renderIcon = (type: string, path: string, name: string) => {
```
Extract out as a component. You may send `classes` as a prop.

rart commented on ui/app/src/components/MediaCard.tsx:
```diff
@@ -119,68 +113,53 @@ const useStyles = makeStyles((theme: Theme) => ({
+    item,
+    previewAppBaseUri,
+    currentView = 'grid',
+    hasCheckbox = true,
```
Can we remove this prop? If the handleSelect prop is not supplied, then there's not checkbox. Also, rename all even handling-props to `on{Action}` (e.g. onEdit, onPreview, onSelect, etc)

rart replied:
shift + f6 (shift+fn+f6) on IntelliJ for assisted rename. 

You'll get something like `handleEdit: onEdit`. Then, manually edit the props to the new name by removing the old name so it ends up as just `onEdit` (~handleEdit:~ onEdit => onEdit)

rart commented on ui/app/src/components/SearchBar.tsx:
```diff
@@ -84,15 +87,15 @@ export default function SearchBar(props: SearchBarProps) {
   const {formatMessage} = useIntl();
   return (
-    <div className={clsx(classes.search, focus && 'focus')}>
+    <div className={clsx(classes.search, focus && 'focus', props.classes?.root)}>
```
Doesn't seem like the focus class has any styles associated to it. Remove?

aachavarria replied:
'&.focus': {
      backgroundColor: palette.white,
      boxShadow: '0px 0px 3px rgba(65, 69, 73, 0.15), 0px 4px 4px rgba(65, 69, 73, 0.15)'
    }

rart commented on ui/app/src/components/SearchBar.tsx:
```diff
@@ -84,15 +87,15 @@ export default function SearchBar(props: SearchBarProps) {
         placeholder={placeholder || formatMessage(messages.placeholder)}
         autoFocus={autofocus}
-        value={keyword}
+        value={typeof keyword !== 'undefined' ? keyword : undefined}
```
When will keyword be undefined?

React will complain as it switches from controlled<=>uncontrolled inputs. 

useState calls that control this component should be initialised with `''` rather than undefined.

rart commented on ui/app/src/models/Card.ts:
```diff
@@ -0,0 +1,24 @@
+/*
```
This name is way too generic for it's purpose. Do even need a separate file for this?

aachavarria replied:
deleted

rart commented on ui/app/src/models/Search.ts:
```diff
@@ -16,12 +16,12 @@
+  query?: string;
   keywords: string;
   offset: number;
   limit: number;
```
Get these by extending from PaginationOptions

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -175,15 +184,19 @@ export function PreviewConcierge(props: any) {
     switch (selectedTool) {
       case 'craftercms.ice.assets':
-        // TODO: aaron to fetch assets here...
+        (assets.isFetching === null && site) && dispatch(fetchPanelAssetsItems(initialSearchParameters));
```
When switching sites, are you resetting the state so the isFetching is null? Test site switching works as expected. The preview toolbar switcher is great for testing this.

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -225,6 +238,15 @@ export function PreviewConcierge(props: any) {
 
 }
 
+const initialSearchParameters: ElasticParams = {
```
Get this from state. Remove from here.

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -175,15 +184,19 @@ export function PreviewConcierge(props: any) {
-        contentTypes && contentTypes$.next(contentTypes);
+        // (!contentTypes && site) && dispatch(fetchContentTypes());
+        // contentTypes && contentTypes$.next(contentTypes);
         break;
```
x

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,232 @@
     defaultMessage: 'Assets'
+  },
+  itemsPerPage: {
+    id: 'search.itemsPerPage',
```
Is this the exact same text and id from the search page?

Make sure they are the same or use a unique id for the Panel

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,232 @@
+  filters: {
+    'mime-type': ['image/png', 'image/jpeg', 'image/gif', 'video/mp4', 'image/svg+xml']
+  }
+};
```
Don't need this. It's on the state.

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,232 @@
+  const {
+    classes,
+    assetsResource,
+    handleChangePage,
```
use `onPageChanged`. (Follow the `on{ACTION}` convention.)

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,232 @@
+        }}
+        nextIconButtonProps={{
+          'aria-label': 'next page',
+        }}
```
Translate

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,232 @@
+      {
+        total === 0 &&
+        <EmptyState title={formatMessage(translations.noResults)}
+                    classes={{ image: classes.noResultsImage, title: classes.noResultsTitle }}/>
```
Format

rart commented on ui/app/src/modules/Search/Search.tsx:
```diff
@@ -266,14 +286,12 @@ function Search(props: SearchProps) {
-                onGetUserPermissions={onGetUserPermissions}
-                mode={mode}
                 previewAppBaseUri={previewAppBaseUri}
+                handleHeaderButtonClick={handleHeaderButtonClick}
```
Rename all `handle{...}` to `on{...}`

rart commented on ui/app/src/modules/Search/Search.tsx:
```diff
@@ -607,6 +657,37 @@ function Search(props: SearchProps) {
         </div>
       </Dialog>
+      <Menu
+        id="options-menu"
```
remove id

rart commented on ui/app/src/services/content.ts:
```diff
@@ -395,8 +395,25 @@ function writeContentUrl(qs: object) {
+        fileName: getInnerHtml(doc.querySelector('file-name'))
+      };
+
+      let fieldNode = doc.querySelector(`:scope > ${fieldId}`);
```
fieldNode may not yet exist. Check for null before setting the innerHTML. If null, you need to first append the new child and then you can set the innerHTML. Basically add an if in between the query and the assignment which body creates the element and appends to the doc.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -410,7 +427,7 @@ export function insertComponent(
   return getDOM(site, modelId).pipe(
     switchMap((doc) => {
-
+      debugger;
```
x

rart commented on ui/app/src/services/search.ts:
```diff
@@ -16,11 +16,14 @@
 
-export function search(siteId: string, parameters: any = {}) {
+export function search(siteId: string, parameters: any = {}): Observable<SearchResult> {
   return post(`/studio/api/2/search/search.json?siteId=${siteId}`, parameters , {
     'Content-Type': 'application/json'
-  })
+  }).pipe(map(({ response }) => response.result))
 }
```
👏🏻

rart commented on ui/app/src/state/actions/preview.ts:
```diff
@@ -175,4 +179,20 @@ export function setItemBeingDragged(active: boolean): StandardAction {
+    type: FETCH_PANEL_ASSETS_ITEMS_FAILED,
+    payload: error
+  };
+}
```
You may use createAction for these too. 

(The mix of usage is mostly due to stuff that was created before we started using redux-toolkit)

rart commented on ui/app/src/state/reducers/preview.ts:
```diff
@@ -227,7 +242,32 @@ const reducer = createReducer<GlobalState['preview']>({
+  }),
+  [FETCH_PANEL_ASSETS_ITEMS_COMPLETE]: (state, { payload: searchResult }: { payload: SearchResult }) => {
+    let itemsLookupTable = createLookupTable<MediaItem>(searchResult.items, 'path');
+    let pageNumber = state.assets.query.offset / state.assets.query.limit;
```
You need to Math.ceil this. Should always be an int.

rart replied:
Check other places you may be doing this op for same issue.

rart commented on ui/app/src/state/reducers/preview.ts:
```diff
@@ -227,7 +242,32 @@ const reducer = createReducer<GlobalState['preview']>({
+  },
+  [FETCH_PANEL_ASSETS_ITEMS_FAILED]: (state, { payload }) => ({
+    ...state,
+    assets: { ...state.assets, error: payload.response, isFetching: true }
```
isFetching should be set to false here

rart commented on ui/app/src/utils/hooks.ts:
```diff
@@ -74,7 +74,7 @@ export function createResource<T = any>(factoryFn: () => Promise<T>) {
   return resource;
```
Remove this file from your commit. There are a few updates that will take place here.

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -697,35 +691,21 @@ export function Guest(props) {
+      if (asset.mimeType.includes('image/')) {
+        type = 'image';
+      } else if (asset.mimeType.includes('video/')) {
+        type = 'video-picker';
```
what's a video-picker type?

aachavarria replied:
is the video asset type, I thought it was 'video' but it is named 'video-picker'

aachavarria committed:
Revert commit

aachavarria committed:
mediaCard Refactor

aachavarria committed:
renaming PanelAssets to AssetsPanel

aachavarria committed:
adding math.ceil

aachavarria committed:
revert hooks.ts

aachavarria committed:
removing models/card, moving CardMenuOption interface

aachavarria committed:
PR Review

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -283,6 +288,26 @@ export function GuestProxy(props) {
         }
+        case UPDATE_FIELD_VALUE_OPERATION:
+          const { modelId, fieldId, value } = op.args;
+          const updatedField = $(`[data-craftercms-model-id="${modelId}"][data-craftercms-field-id="${fieldId}"]`);
```
Repeat group items aren’t being considered

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -283,6 +288,26 @@ export function GuestProxy(props) {
+          const updatedField = $(`[data-craftercms-model-id="${modelId}"][data-craftercms-field-id="${fieldId}"]`);
+          const model = contentController.getCachedModel(modelId);
+          const contentType = contentController.getCachedContentType(model.craftercms.contentType);
+          const assetType = contentType.fields[fieldId].type;
```
This should be called fieldType

rart commented on ui/app/src/components/MediaCard.tsx:
```diff
@@ -265,49 +232,37 @@ function MediaCard(props: MediaCardProps) {
+        {
+          onHeaderButtonClick &&
+          <IconButton
+            aria-label="options"
```
Translation

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,236 @@
+        <React.Suspense
+          fallback={
+            <LoadingState
+              title="Loading"
```
Translation

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,236 @@
+  const assets: PagedEntityState<MediaItem> = assetsResource.read();
+  const { byId, count: total, query, page } = assets;
+  const pageNumber = Math.ceil(query.offset / query.limit);
+  const items = page[pageNumber];
```
make this a pre-parsed prop of the resource. The resource should be nothing more than what the component needs

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,236 @@
+    onDragStart,
+    onDragEnd
+  } = props;
+  const { GUEST_BASE } = useSelector<GlobalState, GlobalState['env']>(state => state.env);
```
prop

rart commented on ui/app/src/services/content.ts:
```diff
@@ -395,8 +395,30 @@ function writeContentUrl(qs: object) {
+
+      let fieldNode = doc.querySelector(`:scope > ${fieldId}`);
+      if (!fieldNode) {
+        fieldNode = document.createElement(fieldId);
```
Though this works, create via `doc` rather than the browser’s document.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -395,8 +395,30 @@ function writeContentUrl(qs: object) {
+      let fieldNode = doc.querySelector(`:scope > ${fieldId}`);
+      if (!fieldNode) {
+        fieldNode = document.createElement(fieldId);
+        doc.append(fieldNode);
```
appendChild

rart commented on ui/app/src/state/reducers/preview.ts:
```diff
@@ -227,7 +242,32 @@ const reducer = createReducer<GlobalState['preview']>({
+  }),
+  [FETCH_ASSETS_PANEL_ITEMS_COMPLETE]: (state, { payload: searchResult }: { payload: SearchResult }) => {
+    let itemsLookupTable = createLookupTable<MediaItem>(searchResult.items, 'path');
+    let pageNumber = Math.ceil(state.assets.query.offset / state.assets.query.limit);
```
We’re having this calculation in many places. I’m thinking we can add the computed value to the state and just use. (i.e. `currentPage`)

aachavarria committed:
fixing util.js setProperty,
adding index to support updateField on repeatable items

aachavarria committed:
update CrafterCMSNextBridge with useLayoutEffect to support setRequestForgeryToken()

aachavarria committed:
adding translations

aachavarria committed:
adding pageNumber to the fetch_assets_panel reducer

aachavarria committed:
creating AssetResource

aachavarria committed:
changing to appendChild

aachavarria committed:
Fixing the catch of ajax errors so that Epics don't throw and get deactivated during runtime

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

 Conflicts:
	ui/app/src/modules/Preview/PreviewConcierge.tsx

aachavarria committed:
uncomment

aachavarria committed:
adding videoThumbnail to mediaCards

aachavarria committed:
video thumbnail width fix

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

aachavarria committed:
adding isElementInView

aachavarria committed:
scroll fixes, UPDATE_FIELD_VALUE_OPERATION changes

aachavarria committed:
GuestProxy.js = revert change,
Guest.js = Fixing issue with array of fieldId,
contentController.js = reordering operations args

aachavarria committed:
Fixes to change parameters ordering

aachavarria committed:
adding Effect dependency 'assets'

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/rart/studio-ui into feature/3632

 Conflicts:
	ui/app/src/models/GlobalState.ts
	ui/app/src/state/actions/preview.ts
	ui/guest/src/components/GuestProxy.js

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

aachavarria committed:
reducer to fix issue on search assets

aachavarria committed:
adding upload from desktop wip

aachavarria committed:
adding XSRF_TOKEN_HEADER_NAME, XSRF_TOKEN to previewConcierge hostToGuest,
dragenter fixes, drag from desktop onDragEnd

aachavarria committed:
adding dnd from desktop to assetpanel

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -85,7 +88,14 @@ export function PreviewConcierge(props: any) {
         case GUEST_CHECK_IN: {
 
-          hostToGuest$.next({ type: HOST_CHECK_IN, payload });
+          hostToGuest$.next({
```
This approach is valid but so far we're only doing this so Guest can do the upload. Since we're working with data URLs, let's change the approach and just have Guest post the data to Host and Host would do the actual upload.

This would render this unnecessary, so for now just leave as `hostToGuest$.next({ type: HOST_CHECK_IN });`

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -159,6 +169,12 @@ export function PreviewConcierge(props: any) {
           break;
         }
         case UPDATE_FIELD_VALUE_OPERATION: {
+          const { modelId, fieldId, index, value } = payload;
```
The children map has been migrated to contentController. You may now send the parent modelId for embedded components.

aachavarria replied:
pending

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,346 @@
+        limit: source.query.limit,
+        pageNumber: source.pageNumber,
+        items
+      }
```
Let's keep the props consistent with the state prop.

```js
    resultSelector: source => {
       const items = source.page[source.pageNumber].map((id) => source.byId[id]);
        return {
          ...pluckProps(source, 'count', 'query.limit', 'pageNumber'),
          items
        }
```

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,346 @@
+      return;
+    }
+
+    const reader = new FileReader();
```
Abstract the upload logic so you can run it both from here and preview concierge for the two different use cases.

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,346 @@
+  document.addEventListener("dragleave", function (event) {
+    //needs to know if it left the page
+    //setDragInProgress(false);
+  }, false);
```
1. Check your code style settings to set single quote as the quote setting.
2. This things are side effects; they should be inside a useEfffect 😱 
3. It could be something like this:
```ts
// import fromEvent from rxjs
// ...
const elementRef = useRef();
// ...
useEffect(() => {
  const subscription = fromEvent(elementRef.current, 'dragenter').pipe(
    filter(e => e.dataTransfer?.types.includes('Files'))
  ).subscribe(() => {
    setDragInProgress(true);
  });
  return () => subscription.unsubscribe();
}, []);
useEffect(() => {
  if (dragInProgress) {
    const subscription = fromEvent(elementRef.current, 'dragend').subscribe(() => {
      setDragInProgress(false);
      // whatever else you need to do...
    });
    return () => subscription.unsubscribe();
  }
}, [dragInProgress]);
// ...
return (
  <RootElement ref={elementRef}...
```

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,346 @@
+                onDragStart={onDragStart}
+                onDragEnd={onDragEnd}
+              />
+            )
```
Use simplified return syntax here.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -398,8 +398,43 @@ function writeContentUrl(qs: object) {
 
-export function updateField(modelId: string, fieldId: string, value: any): Observable<any> {
-  throw new Error('Not implemented.');
+export function updateField(site: string, modelId: string, fieldId: string, index: number, value: any): Observable<any> {
```
Once you merge the latest updates, add the call to `updateModifiedDateElement(doc);`

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -50,6 +50,31 @@ export function getPreviewURLFromPath(baseUrl: string, path: string) {
   return `${baseUrl}${url}`;
 }
 
+export function dataURItoBlob(dataURI) {
```
- Move to string utils.
- Type the parameter.
- Rename to dataUriToBlob

rart commented on ui/guest/package.json:
```diff
@@ -20,6 +20,7 @@
     "react-dom": "^16.12.0",
     "react-scripts": "3.3.0-next.62",
     "rxjs": "^6.5.3",
+    "uppy": "^1.7.0",
```
As mentioned, let's do the update Studio-side. So you may `yarn remove` this change (rollback).

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -51,6 +54,9 @@ import { DropMarker } from './DropMarker';
 import Cookies from 'js-cookie';
+//import { XHRUpload, Core } from 'uppy';
+import Core from '@uppy/core';
+import XHRUpload from '@uppy/xhr-upload';
```
Remove

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -659,6 +657,48 @@ export function Guest(props) {
           // TODO: Insert detached component
           break;
         }
+        case EditingStatus.UPLOAD_ASSET_FROM_DESKTOP: {
```
Update to perform the actual upload studio-side

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -978,8 +1004,76 @@ export function Guest(props) {
+    },
+
+    dragenter(e, record) {
+      if (stateRef.current.common.status === EditingStatus.LISTENING && e.originalEvent.dataTransfer.types.includes('Files')) {
```
I think dataTransfer or types may be null/undefined and throw.

`e.originalEvent.dataTransfer?.types.?includes('...`

Break the if conditions each on it's own line.

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -1007,14 +1101,34 @@ export function Guest(props) {
     }
   }
 
+  function scrollToReceptacle(receptacles) {
+    let elementInView;
+    let element;
+    elementInView = forEach(receptacles, ({ id }) => {
+      let elem = ElementRegistry.fromICEId(id).element;
+      if (isElementInView(elem)) {
+        elementInView = true;
+        element = elem;
+        return 'break';
+      }
+    }, false);
+
+    if (!elementInView) {
+      // TODO: Do this relative to the scroll position. Don't move if things are already in viewport. Be smarter.
+      let element = ElementRegistry.fromICEId(receptacles[0].id).element;
+      $(scrollElement).animate({
+        scrollTop: $(element).offset().top - 100
+      }, 300);
+    }
+  }
```
Is the bug we saw fixed?

aachavarria replied:
no, it is just the function scrollToReceptacle with changes to use the receptacles object

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -1086,8 +1200,9 @@ export function Guest(props) {
     }, 700);
 
-    const hostDetectionSubscription = fromTopic(HOST_CHECK_IN).pipe(take(1)).subscribe(() => {
+    const hostDetectionSubscription = fromTopic(HOST_CHECK_IN).pipe(take(1)).subscribe((payload) => {
```
Not needed anymore (upload will be done studio-side)

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -285,6 +293,26 @@ export function GuestProxy(props) {
+
+          if (fieldType === 'image') {
+            const tagName = updatedField.eq(index).prop('tagName');
+            if (tagName === 'IMG') {
```
Let's do `.prop(tagName).toLowerCase()` and `if tagName === img`

rart commented on ui/guest/src/util.js:
```diff
@@ -498,3 +500,40 @@ export function addClickListener(element, type, handler) {
+  }
+}
+
+export function dataURItoBlob(dataURI) {
```
Name dataUriToBlob

rart replied:
Can we remove this function from here? It's no longer used guest-side, is it?

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/rart/studio-ui into feature/3632

 Conflicts:
	ui/app/src/modules/Preview/PreviewConcierge.tsx
	ui/guest/src/components/Guest.js

aachavarria committed:
dnd upload on assetspanel

aachavarria committed:
adding dataURItoBlob to string.ts,
moving guest upload asset to previewConcierge

aachavarria committed:
adding uploadDataUrl service

aachavarria committed:
PR review changes

aachavarria committed:
assetPanel/toolpanel: adding new style to hide the scrollbar,
Guest: adding document listener to dragenter and dragleave

aachavarria committed:
fixing conditional

rart commented on ui/app/src/modules/Preview/Tools/ToolPanel.tsx:
```diff
@@ -76,7 +76,7 @@ export function ToolPanel(props: ToolPanelProps): ReactElement | null {
   } = props;
   return (
-    <div className={classes}>
+    <div className={classes? classes: ''}>
```
You can do `classes ?? ''`. 

However, I would think `className={classes}` would be ok as react would handle the nullish value, no? Why did you had to do this change?

aachavarria replied:
fixed

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -234,7 +238,7 @@ export default function AssetsPanel() {
+      ).subscribe(() => {
         clearTimeout(timeoutRef.current);
         timeoutRef.current = setTimeout(() => {
           clearTimeout(timeoutRef.current);
```
Let's give this a try....
```js
useEffect(() => {
  if (dragInProgress) {
    const dragover$ = fromEvent(elementRef.current, 'dragover').pipe(
      tap((e) => {
        e.preventDefault();
        e.stopPropagation();
      })
    );
    const subscription = fromEvent(elementRef.current, 'dragleave').pipe(
      switchMap(() => interval(100).pipe(takeUntil(dragover$))),
      mapTo(false)
    ).subscribe(setDragInProgress);
    return () => {
      subscription.unsubscribe();
    };
  }
}, [dragInProgress]);
```

aachavarria replied:
it is working but has a issue with the dragover

rart commented on ui/app/src/modules/Preview/Tools/ToolPanel.tsx:
```diff
@@ -71,10 +72,11 @@ export function ToolPanel(props: ToolPanelProps): ReactElement | null {
   } = props;
   return (
-    <>
+    <div className={classes}>
```
Do you really need to wrap the whole ToolPanel? Wouldn't it suffice with applying the class to the body (i.e. the section element)?

rart commented on ui/app/src/modules/Preview/Tools/ToolPanel.tsx:
```diff
@@ -38,13 +38,14 @@ const useStyles = makeStyles((theme: Theme) => createStyles({
   BackIcon?: ElementType,
-  onBack?: () => void
+  onBack?: () => void,
+  classes?: any
```
Classes, as in other components should be an object with the classes for each piece of the ui. e.g. `{ header, body }`; type it.

rart commented on ui/app/src/modules/Preview/Tools/ToolPanel.tsx:
```diff
@@ -38,13 +38,14 @@ const useStyles = makeStyles((theme: Theme) => createStyles({
   title: string;
   BackIcon?: ElementType,
-  onBack: () => void
+  onBack: () => void,
```
Check your code style settings for trailing comma to be disabled!

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -1212,6 +1199,27 @@ export function Guest(props) {
+        clearTimeout(persistence.dragLeaveTimeout);
+        fn.onDragEnd();
+      }, 100);
+    });
```
1. You're missing the destruction logic
2. Try the rxjs-way that I added on a prior comment for the AssetsPanel dragenter/dragleave/dragover

aachavarria replied:
pending

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -84,7 +84,6 @@ export function GuestProxy(props) {
       .on('dragstart', '[data-craftercms-model-id]', handler)
-      .on('dragenter', '[data-craftercms-model-id]', handler)
       .on('dragover', '[data-craftercms-model-id]', handler)
       .on('dragleave', '[data-craftercms-model-id]', handler)
```
Do we need dragleave?

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -227,39 +223,30 @@ export function PreviewConcierge(props: any) {
+              name: payload.name,
+              type: payload.type,
+              dataUrl: payload.dataUrl,
+            },
```
pluckProps(payload, 'name', 'type', 'dataUrl')

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -198,27 +195,25 @@ export default function AssetsPanel() {
+          name: file.name,
+          type: file.type,
+          dataUrl: reader.result,
+        },
```
pluckProps(...)

aachavarria committed:
Merge branch 'feature/3632' of https://github.com/yureii00/studio-ui into feature/3632

 Conflicts:
	ui/guest/src/components/Guest.js

aachavarria committed:
fixing scroll issue on ToolPanel,
changing useEffect for dragleave and dragover on AssetsPanel

aachavarria committed:
adding pluckProps,

aachavarria committed:
revert to old effect, the new one is not getting fine the e.preventDefault(), e.stopPropagation();
the drop event is still going to the asset

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -15,24 +15,382 @@
+  //       subscription.unsubscribe();
+  //     };
+  //   }
+  // }, [dragInProgress]);
```
What's up with this?

aachavarria replied:
lets have a call

rart commented on ui/app/src/modules/Preview/Tools/ToolPanel.tsx:
```diff
@@ -38,7 +38,10 @@ const useStyles = makeStyles((theme: Theme) => createStyles({
-  onBack?: () => void
+  onBack?: () => void,
+  classes?: {
+    body: any;
```
body should be optional too

rart commented on ui/app/src/utils/string.ts:
```diff
@@ -69,3 +69,27 @@ export default {
   formatBytes
 };
+
+export function dataUriToBlob(dataURI) {
```
type missing

rart commented:
Please merge latest from 2019 branch from upstream so the changes you pulled from my fork don't show as changes of your PR.

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

aachavarria committed:
removing not used function

aachavarria committed:
toolpanel: additional parameter,
string.ts: Typo

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3632

 Conflicts:
	ui/app/src/modules/Preview/PreviewConcierge.tsx
	ui/app/src/services/content.ts
	ui/guest/src/util.js

aachavarria committed:
updating effect to control dnd

aachavarria committed:
updates guest effect to handle dnd operations

aachavarria committed:
adding parentModelId to updateField,
modifying content.ts updateField to support embeddedComponents

aachavarria committed:
adding progress support to upload asset

aachavarria committed:
removing console.log

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -240,10 +277,9 @@ export function PreviewConcierge(props: any) {
     switch (selectedTool) {
       case 'craftercms.ice.assets':
-        // TODO: aaron to fetch assets here...
+        (assets.isFetching === null && site) && dispatch(fetchAssetsPanelItems(assets.query));
```
Needs to consider that there’s no error in this condition

rart replied:
What I meant was the fetch/load condition needs to consider error. Otherwise, it goes in a loop of fetching. The feedback display comes in the form of ErrorBoundary at the panel level so no need to display the snack.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -398,8 +401,43 @@ function writeContentUrl(qs: object) {
+    parentModelId,
+    doc => {
+      let fieldNode;
+      if (typeof indexToUpdate === 'string') {
```
This logic is not sufficient for all cases of update.

For example, see header content type. Scenario: updating social_media.icon[0]. You should probably use extractNode in either case. And it may need a little tweaking to support.

rart commented on ui/guest/src/classes/ContentController.js:
```diff
@@ -173,11 +172,19 @@ export class ContentController {
+      fieldId,
+      index,
+      value,
+      parentModelId: isNullOrUndefined(ModelHelper.prop(model, 'path'))
```
This is repetitive. We should encapsulate and reuse.

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -971,8 +980,68 @@ export function Guest(props) {
+          dropZone.origin = false;
+          dropZones.push(dropZone);
+
+          siblings = [...siblings, ...dropZone.children];
```
I don’t think siblings or children are applicable to asset drop zones.

rart reviewed (changes requested):
Check your code style settings. Set to always add semicolons and set continuation to 2 spaces as well.

See comments below

Check unresolved above (open hidden conversations).

Once you’re done with these, remove draft status.

aachavarria committed:
encapsulating getParentModelId logic

aachavarria committed:
adding error conditional

aachavarria committed:
removing childrens, and siblings

aachavarria committed:
adding index null case to extractNode,

rart committed:
Ruggedizing condition

aachavarria committed:
fixing conditional

aachavarria committed:
Merge branch 'feature/3632' of https://github.com/yureii00/studio-ui into feature/3632

rart commented on ui/guest/src/classes/ContentController.js:
```diff
@@ -544,6 +547,12 @@ export class ContentController {
 
 }
 
+function getParentModelId(modelId, model, models, children) {
```
`model` can be retrieved by `models[modelId]` without additional costs

aachavarria committed:
getParentModelId changes

aachavarria committed:
removing siblings, and childrens from assetDragStarted

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -648,6 +639,41 @@ export function Guest(props) {
+                    payload.path
+                  )
+                }
+              });
```
```js
              message$.pipe(
                filter((e) => 
                  (e.data?.type === DESKTOP_ASSET_UPLOAD_COMPLETE) &&
                  (e.data.id === file.name)
                ),
                map(e => e.data),
                take(1)
              ).subscribe(function () {
                  contentController.updateField(
                    record.modelId,
                    record.fieldId[0],
                    record.index,
                    payload.path
                  )
              });
```

aachavarria committed:
updating updateField logic

aachavarria committed:
PR review

rart committed:
Removed console log
